src/azure/credentials.rs

```diff
@@ -0,0 +1,184 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env::*;
+
+use errors::*;
+
+#[derive(Clone, Debug)]
+pub struct AzureCredentials {
+    blob_endpoint: String,
+    account_name: String,
+    account_key: String,
+    container_name: Option<String>,
+}
+
+impl AzureCredentials {
+    pub fn new(blob_endpoint: &str, account_name: &str, account_key: &str, container_name: Option<String>) -> AzureCredentials {
+
+        let endpoint = if blob_endpoint.ends_with("/") {
+            blob_endpoint.to_owned()
+        } else {
+            blob_endpoint.to_owned() + "/"
+        };
+
+        AzureCredentials {
+            blob_endpoint: endpoint,
+            account_name: account_name.to_owned(),
+            account_key: account_key.to_owned(),
+            container_name: container_name,
+        }
+    }
+
+    pub fn azure_blob_endpoint(&self) -> &str {
+        &self.blob_endpoint
+    }
+
+    pub fn azure_account_name(&self) -> &str {
+        &self.account_name
+    }
+
+    pub fn azure_account_key(&self) -> &str {
+        &self.account_key
+    }
+
+    pub fn blob_container_name(&self) -> &Option<String> {
+        &self.container_name
+    }
+}
+
+pub trait AzureCredentialsProvider {
+    fn provide_credentials(&self) -> Result<AzureCredentials>;
+}
+
+pub struct EnvironmentProvider;
+
+impl AzureCredentialsProvider for EnvironmentProvider {
+    fn provide_credentials(&self) -> Result<AzureCredentials> {
+        credentials_from_environment()
+    }
+}
+
+fn credentials_from_environment() -> Result<AzureCredentials> {
+    let env_conn_str = var("SCCACHE_AZURE_CONNECTION_STRING").chain_err(|| {
+        "No SCCACHE_AZURE_CONNECTION_STRING in environment"
+    })?;
+
+    let container_name = match var("SCCACHE_AZURE_BLOB_CONTAINER") {
+        Ok(text) => {
+            if text.is_empty() {
+                None
+            } else {
+                Some(text)
+            }
+        }
+
+        Err(_) => None
+    };
+
+    parse_connection_string(&env_conn_str, container_name)
+}
+
+fn parse_connection_string(conn: &str, container_name: Option<String>) -> Result<AzureCredentials> {
+    let mut blob_endpoint = String::default();
+    let mut default_endpoint_protocol: String = "https".to_owned();
+    let mut account_name = String::default();
+    let mut account_key = String::default();
+    let mut endpoint_suffix = String::default();
+
+    let split = conn.split(";");
+    for part in split {
+        if part.starts_with("BlobEndpoint=") {
+            blob_endpoint = substr(part, "BlobEndpoint=".len()).to_owned();
+            continue;
+        }
+
+        if part.starts_with("DefaultEndpointsProtocol=") {
+            default_endpoint_protocol = substr(part, "DefaultEndpointsProtocol=".len()).to_owned();
```

Comment:
I thought there was a better way to do this, but I don't think there is. I believe what I was thinking of was the nice [`split_prefix` function](https://github.com/mozilla/sccache/blob/7079ab87832efab4298757176b1388f1d126ed98/src/util.rs#L171) we've got for `OsStr`.